signing.go

```diff
@@ -0,0 +1,105 @@
+package matrixfederation
+
+import (
+	"encoding/json"
+	"fmt"
+	"golang.org/x/crypto/ed25519"
+)
+
+// SignJSON signs a JSON object returning a copy signed with the given key.
+// https://matrix.org/docs/spec/server_server/unstable.html#signing-json
+func SignJSON(entityName, keyID string, privateKey ed25519.PrivateKey, message []byte) ([]byte, error) {
+	var object map[string]*json.RawMessage
+	var signatures map[string]map[string]Base64String
+	if err := json.Unmarshal(message, &object); err != nil {
```

Comment:
Yes we are mostly signing ``structs``. But turning a struct into ``map[string]*json.RawMessage`` is a faff, and leaks implementation details of the JSON signing into the rest of the code.

```
bytes1, err := json.Marshal(&mystruct.field1)
bytes2, err := json.Marshal(&mystruct.field2)
signed, err := SignJSON(...,map[string]*json.RawMessage{"field1": bytes1, "field2": bytes2})
```
vs
```
bytes, err := json.Marshal(&mystruct)
signed, err := SignJSON(...,bytes)
```